app/controllers/application_controller.rb

```diff
@@ -1,2 +1,34 @@
 class ApplicationController < ActionController::Base
+  private
+
+  def authenticate_user!
+    redirect_to root_path, alert: "You must be logged in to do that." unless user_signed_in?
+  end
+
+  def current_user
+    Current.user ||= authenticate_user_from_session
+  end
+
+  helper_method :current_user
+
+  def authenticate_user_from_session
+    User.find_by(id: session[:user_id])
+  end
+
+  def user_signed_in?
+    current_user.present?
+  end
+
+  helper_method :user_signed_in?
+
+  def login(user)
```

Comment:
This should be in the sessions controller instead. ApplicationController is a base for all controllers in the app which should only include methods where most controllers will use. Login and logout methods should be use in the sessions controller instead. !LGTM